File: iosApp/iosApp/repository/FirestoreFeedRepository.swift
import Foundation
import os.log
import FirebaseFirestore
import Shared

/// Firestore implementation of FeedRepository.
/// Reads from users/{currentUserId}/feed (fan-out by onPostCreate Cloud Function).
final class FirestoreFeedRepository: FeedRepository {

    private static let log = Logger(subsystem: "com.kissangram", category: "FirestoreFeedRepo")
    private let firestore = Firestore.firestore(database: "kissangram")
    private let authRepository: AuthRepository
    private var lastDocumentSnapshot: DocumentSnapshot?
    private var lastPublicPostsSnapshot: DocumentSnapshot?
    private var isFallbackMode = false
    private var cachedPosts: [Post]?

    init(authRepository: AuthRepository) {
        self.authRepository = authRepository
    }

    func getHomeFeed(page: Int32, pageSize: Int32, forceRefresh: Bool) async throws -> [Post] {
        Self.log.debug("getHomeFeed: page=\(page) pageSize=\(pageSize)")
        guard let currentUserId = try await authRepository.getCurrentUserId() else {
            Self.log.warning("getHomeFeed: currentUserId is nil, returning empty")
            return []
        }
        let uidPrefix = String(currentUserId.prefix(8))
        Self.log.debug("getHomeFeed: currentUserId=\(uidPrefix)...")
        guard page >= 0 else { return [] }
        guard pageSize >= 1, pageSize <= 50 else { return [] }

        // Return cached page 0 when not forcing refresh (avoids reload on tab return)
        if page == 0 && !forceRefresh, let cached = cachedPosts {
            Self.log.debug("getHomeFeed: returning cached posts (\(cached.count))")
            return cached
        }

        if forceRefresh {
            lastDocumentSnapshot = nil
            lastPublicPostsSnapshot = nil
            isFallbackMode = false
        }

        let postsCollection = firestore.collection("posts")

        // When not in fallback mode, try feed first
        if !isFallbackMode {
            let feedRef = firestore.collection("users").document(currentUserId).collection("feed")
            var feedQuery: Query = feedRef.order(by: "createdAt", descending: true).limit(to: Int(pageSize))

            if page > 0 {
                guard let lastSnap = lastDocumentSnapshot else {
                    Self.log.warning("getHomeFeed: page>0 but no cursor, returning empty")
                    return []
                }
                feedQuery = feedQuery.start(afterDocument: lastSnap)
                Self.log.debug("getHomeFeed: using startAfter for page \(page)")
            } else {
                lastDocumentSnapshot = nil
            }

            Self.log.debug("getHomeFeed: executing query users/\(uidPrefix).../feed orderBy createdAt desc limit \(pageSize)")
            let snapshot = try await feedQuery.getDocuments()
            let rawCount = snapshot.documents.count
            Self.log.debug("getHomeFeed: snapshot.documents.count=\(rawCount)")

            let postsWithIds = snapshot.documents.compactMap { doc -> Post? in
                let post = toPost(from: doc, isLikedByMe: false)
                if post == nil { Self.log.warning("getHomeFeed: toPost returned nil for doc \(doc.documentID)") }
                return post
            }

            // If feed has content, use it
            if !postsWithIds.isEmpty {
                let likedPostIds = try await batchCheckLikes(postIds: postsWithIds.map { $0.id }, userId: currentUserId)
                let likesCountByPostId = try await batchFetchLikesCount(postIds: postsWithIds.map { $0.id })
                let posts = postsWithIds.map { post in
                    let freshLikesCount = likesCountByPostId[post.id] ?? post.likesCount
                    return Post(
                        id: post.id,
                        authorId: post.authorId,
                        authorName: post.authorName,
                        authorUsername: post.authorUsername,
                        authorProfileImageUrl: post.authorProfileImageUrl,
                        authorRole: post.authorRole,
                        authorVerificationStatus: post.authorVerificationStatus,
                        type: post.type,
                        text: post.text,
                        media: post.media,
                        voiceCaption: post.voiceCaption,
                        crops: post.crops,
                        hashtags: post.hashtags,
                        location: post.location,
                        question: post.question,
                        likesCount: freshLikesCount,
                        commentsCount: post.commentsCount,
                        savesCount: post.savesCount,
                        isLikedByMe: likedPostIds.contains(post.id),
                        isSavedByMe: post.isSavedByMe,
                        createdAt: post.createdAt,
                        updatedAt: post.updatedAt
                    )
                }
                Self.log.debug("getHomeFeed: parsed list.count=\(posts.count) raw=\(rawCount)")
                if let last = snapshot.documents.last {
                    lastDocumentSnapshot = last
                }
                if page == 0 {
                    cachedPosts = posts
                }
                return posts
            }

            // Feed empty on page 0: fall back to public posts
            if page == 0 {
                isFallbackMode = true
                lastPublicPostsSnapshot = nil
                Self.log.debug("getHomeFeed: feed empty, falling back to public posts")
            } else {
                Self.log.warning("getHomeFeed: feed empty on page>0; returning empty")
                return []
            }
        }

        // Fallback mode: fetch paginated public posts from posts collection
        var publicQuery: Query = postsCollection
            .whereField("visibility", isEqualTo: "public")
            .whereField("isActive", isEqualTo: true)
            .order(by: "createdAt", descending: true)
            .limit(to: Int(pageSize))

        if page > 0 {
            guard let lastSnap = lastPublicPostsSnapshot else {
                Self.log.warning("getHomeFeed: fallback page>0 but no cursor, returning empty")
                return []
            }
            publicQuery = publicQuery.start(afterDocument: lastSnap)
            Self.log.debug("getHomeFeed: fallback using startAfter for page \(page)")
        } else {
            lastPublicPostsSnapshot = nil
        }

        Self.log.debug("getHomeFeed: fallback query posts visibility=public isActive=true orderBy createdAt desc limit \(pageSize)")
        let snapshot = try await publicQuery.getDocuments()
        let rawCount = snapshot.documents.count
        Self.log.debug("getHomeFeed: fallback snapshot.documents.count=\(rawCount)")

        let postsWithIds = snapshot.documents.compactMap { doc -> Post? in
            let post = toPost(from: doc, isLikedByMe: false)
            if post == nil { Self.log.warning("getHomeFeed: toPost returned nil for doc \(doc.documentID)") }
            return post
        }

        let likedPostIds = if !postsWithIds.isEmpty {
            try await batchCheckLikes(postIds: postsWithIds.map { $0.id }, userId: currentUserId)
        } else {
            Set<String>()
        }
        let likesCountByPostId = if !postsWithIds.isEmpty {
            try await batchFetchLikesCount(postIds: postsWithIds.map { $0.id })
        } else {
            [String: Int32]()
        }

        let posts = postsWithIds.map { post in
            let freshLikesCount = likesCountByPostId[post.id] ?? post.likesCount
            return Post(
                id: post.id,
                authorId: post.authorId,
                authorName: post.authorName,
                authorUsername: post.authorUsername,
                authorProfileImageUrl: post.authorProfileImageUrl,
                authorRole: post.authorRole,
                authorVerificationStatus: post.authorVerificationStatus,
                type: post.type,
                text: post.text,
                media: post.media,
                voiceCaption: post.voiceCaption,
                crops: post.crops,
                hashtags: post.hashtags,
                location: post.location,
                question: post.question,
                likesCount: freshLikesCount,
                commentsCount: post.commentsCount,
                savesCount: post.savesCount,
                isLikedByMe: likedPostIds.contains(post.id),
                isSavedByMe: post.isSavedByMe,
                createdAt: post.createdAt,
                updatedAt: post.updatedAt
            )
        }

        Self.log.debug("getHomeFeed: fallback parsed list.count=\(posts.count) raw=\(rawCount)")
        if let last = snapshot.documents.last {
            lastPublicPostsSnapshot = last
        }
        if page == 0 {
            cachedPosts = posts
        }
        return posts
    }

    func refreshFeed() async throws -> [Post] {
        Self.log.debug("refreshFeed: clearing cursor, forcing refresh")
        lastDocumentSnapshot = nil
        return try await getHomeFeed(page: 0, pageSize: 20, forceRefresh: true)
    }
    
    // MARK: - Batch Fetch Likes Count
    
    /// Batch-fetch likesCount from posts collection (source of truth).
    /// Feed documents have stale likesCount; posts collection is updated by Cloud Functions on like/unlike.
    private func batchFetchLikesCount(postIds: [String]) async throws -> [String: Int32] {
        if postIds.isEmpty { return [:] }
        return try await withThrowingTaskGroup(of: (String, Int32).self) { group in
            for postId in postIds {
                group.addTask {
                    let doc = try await self.firestore.collection("posts").document(postId).getDocument()
                    if doc.exists, let data = doc.data() {
                        let count = Int32((data["likesCount"] as? Int) ?? 0)
                        return (postId, count)
                    }
                    return (postId, 0)
                }
            }
            var result = [String: Int32]()
            for try await (postId, count) in group {
                result[postId] = count
            }
            return result
        }
    }
    
    // MARK: - Batch Check Likes
    
    /// Batch check which posts are liked by the current user.
    /// Uses parallel async/await for efficient batch fetching.
    private func batchCheckLikes(postIds: [String], userId: String) async throws -> Set<String> {
        if postIds.isEmpty { return Set<String>() }
        
        return try await withThrowingTaskGroup(of: (String, Bool).self) { group in
            var likedPostIds = Set<String>()
            
            // Create tasks to check each post
            for postId in postIds {
                group.addTask {
                    let likeDoc = try await self.firestore
                        .collection("posts")
                        .document(postId)
                        .collection("likes")
                        .document(userId)
                        .getDocument()
                    return (postId, likeDoc.exists)
                }
            }
            
            // Collect results
            for try await (postId, exists) in group {
                if exists {
                    likedPostIds.insert(postId)
                }
            }
            
            Self.log.debug("batchCheckLikes: checked \(postIds.count) posts, found \(likedPostIds.count) liked")
            return likedPostIds
        }
    }

    // MARK: - toPost (same shape as post document in feed)

    private func toPost(from doc: DocumentSnapshot, isLikedByMe: Bool = false) -> Post? {
        guard let data = doc.data() else { return nil }

        let id = (doc.get("id") as? String) ?? doc.documentID
            let authorId = (doc.get("authorId") as? String) ?? ""
            let authorName = (doc.get("authorName") as? String) ?? ""
            let authorUsername = (doc.get("authorUsername") as? String) ?? ""
            let authorProfileImageUrlRaw = doc.get("authorProfileImageUrl") as? String
            let authorProfileImageUrl = authorProfileImageUrlRaw.flatMap { $0.isEmpty ? nil : $0 }
            let authorRoleStr = (doc.get("authorRole") as? String) ?? "farmer"
            let authorVerificationStatusStr = (doc.get("authorVerificationStatus") as? String) ?? "unverified"

            let typeStr = (doc.get("type") as? String) ?? "normal"
            let text = (doc.get("text") as? String) ?? ""

            let mediaArray = (data["media"] as? [[String: Any]])?.compactMap { mediaMap -> PostMedia? in
                guard let url = mediaMap["url"] as? String else { return nil }
                let typeStr = mediaMap["type"] as? String ?? "image"
                let thumbnailUrl = mediaMap["thumbnailUrl"] as? String
                return PostMedia(
                    url: url,
                    type: typeStr == "video" ? .video : .image,
                    thumbnailUrl: thumbnailUrl?.isEmpty == false ? thumbnailUrl : nil
                )
            } ?? []

            let voiceCaptionMap = data["voiceCaption"] as? [String: Any]
            let voiceCaption: VoiceContent? = voiceCaptionMap.flatMap { map in
                guard let url = map["url"] as? String else { return nil }
                let duration = (map["durationSeconds"] as? Int) ?? 0
                return VoiceContent(url: url, durationSeconds: Int32(duration))
            }

            let crops = (data["crops"] as? [String]) ?? []
            let hashtags = (data["hashtags"] as? [String]) ?? []

            let locationMap = data["location"] as? [String: Any]
            let location: PostLocation? = locationMap.flatMap { map in
                let name = map["name"] as? String ?? ""
                let geoPoint = map["geoPoint"] as? GeoPoint
                let latitude = geoPoint?.latitude
                let longitude = geoPoint?.longitude
                return PostLocation(
                    name: name,
                    latitude: latitude != nil ? KotlinDouble(value: latitude!) : nil,
                    longitude: longitude != nil ? KotlinDouble(value: longitude!) : nil
                )
            }

            let questionMap = data["question"] as? [String: Any]
            let question: QuestionData? = questionMap.flatMap { map in
                let targetExpertise = (map["targetExpertise"] as? [String]) ?? []
                let targetExpertIds = (map["targetExpertIds"] as? [String]) ?? []
                let targetExperts = (map["targetExperts"] as? [[String: Any]])?.compactMap { expertMap -> UserInfo? in
                    guard let id = expertMap["id"] as? String,
                          let name = expertMap["name"] as? String,
                          let username = expertMap["username"] as? String else { return nil }
                    let profileImageUrl = expertMap["profileImageUrl"] as? String
                    let roleStr = expertMap["role"] as? String ?? "farmer"
                    let statusStr = expertMap["verificationStatus"] as? String ?? "unverified"
                    return UserInfo(
                        id: id,
                        name: name,
                        username: username,
                        profileImageUrl: profileImageUrl,
                        role: stringToUserRole(roleStr),
                        verificationStatus: stringToVerificationStatus(statusStr)
                    )
                } ?? []
                return QuestionData(
                    targetExpertise: targetExpertise,
                    targetExpertIds: targetExpertIds,
                    targetExperts: targetExperts,
                    isAnswered: (map["isAnswered"] as? Bool) ?? false,
                    bestAnswerCommentId: map["bestAnswerCommentId"] as? String
                )
            }

            let likesCount = (data["likesCount"] as? Int) ?? 0
            let commentsCount = (data["commentsCount"] as? Int) ?? 0
            let savesCount = (data["savesCount"] as? Int) ?? 0

            let createdAt: Int64
            if let createdAtTimestamp = doc.get("createdAt") as? Timestamp {
                createdAt = Int64(createdAtTimestamp.dateValue().timeIntervalSince1970 * 1000)
            } else if let created = data["createdAt"] as? Int64 {
                createdAt = created
            } else {
                createdAt = Int64(Date().timeIntervalSince1970 * 1000)
            }

            let updatedAt: KotlinLong?
            if let updatedAtTimestamp = doc.get("updatedAt") as? Timestamp {
                updatedAt = KotlinLong(value: Int64(updatedAtTimestamp.dateValue().timeIntervalSince1970 * 1000))
            } else if let updated = data["updatedAt"] as? Int64 {
                updatedAt = KotlinLong(value: updated)
            } else {
                updatedAt = nil
            }

            return Post(
                id: id,
                authorId: authorId,
                authorName: authorName,
                authorUsername: authorUsername,
                authorProfileImageUrl: authorProfileImageUrl,
                authorRole: stringToUserRole(authorRoleStr),
                authorVerificationStatus: stringToVerificationStatus(authorVerificationStatusStr),
                type: typeStr == "question" ? .question : .normal,
                text: text,
                media: mediaArray,
                voiceCaption: voiceCaption,
                crops: crops,
                hashtags: hashtags,
                location: location,
                question: question,
                likesCount: Int32(likesCount),
                commentsCount: Int32(commentsCount),
                savesCount: Int32(savesCount),
                isLikedByMe: isLikedByMe,
                isSavedByMe: false, // TODO: Check saved posts similarly
                createdAt: Int64(createdAt),
                updatedAt: updatedAt.map { KotlinLong(value: Int64($0)) }
            )
    }

    private func stringToUserRole(_ roleStr: String) -> UserRole {
        switch roleStr {
        case "expert": return .expert
        case "agripreneur": return .agripreneur
        case "input_seller": return .inputSeller
        case "agri_lover": return .agriLover
        default: return .farmer
        }
    }

    private func stringToVerificationStatus(_ statusStr: String) -> VerificationStatus {
        switch statusStr {
        case "pending": return .pending
        case "verified": return .verified
        case "rejected": return .rejected
        default: return .unverified
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/kissangram/repository/FirestoreFeedRepository.kt
package com.kissangram.repository

import android.util.Log
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.GeoPoint
import com.google.firebase.firestore.Query
import com.kissangram.model.*
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

/**
 * Firestore implementation of FeedRepository.
 * Reads from users/{currentUserId}/feed (fan-out by onPostCreate Cloud Function).
 * Uses cursor-based pagination; stateful for load-more (stores last snapshot).
 */
class FirestoreFeedRepository(
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance(
        com.google.firebase.FirebaseApp.getInstance(),
        DATABASE_NAME
    ),
    private val authRepository: AuthRepository
) : FeedRepository {

    private var lastDocumentSnapshot: DocumentSnapshot? = null

    override suspend fun getHomeFeed(page: Int, pageSize: Int): List<Post> {
        Log.d(TAG, "getHomeFeed: page=$page pageSize=$pageSize")
        val currentUserId = authRepository.getCurrentUserId()
        if (currentUserId == null) {
            Log.w(TAG, "getHomeFeed: currentUserId is null, returning empty list")
            return emptyList()
        }
        Log.d(TAG, "getHomeFeed: currentUserId=${currentUserId.take(8)}...")
        require(page >= 0) { "Page must be non-negative" }
        require(pageSize in 1..50) { "Page size must be between 1 and 50" }

        return try {
            var query: Query = firestore
                .collection("users")
                .document(currentUserId)
                .collection("feed")
                .orderBy("createdAt", Query.Direction.DESCENDING)
                .limit(pageSize.toLong())

            if (page > 0) {
                val lastSnap = lastDocumentSnapshot
                if (lastSnap == null) {
                    Log.w(TAG, "getHomeFeed: page > 0 but no cursor; returning empty")
                    return emptyList()
                }
                query = query.startAfter(lastSnap)
                Log.d(TAG, "getHomeFeed: using startAfter for page $page")
            } else {
                lastDocumentSnapshot = null
            }

            Log.d(TAG, "getHomeFeed: executing query users/$currentUserId/feed orderBy createdAt desc limit $pageSize")
            val snapshot = query.get().await()
            val rawCount = snapshot.documents.size
            Log.d(TAG, "getHomeFeed: snapshot.documents.size=$rawCount")
            
            // First, parse all posts to get their IDs
            val postsWithIds = snapshot.documents.mapNotNull { doc ->
                val post = doc.toPost(isLikedByMe = false) // Temporary, will update after checking likes
                if (post == null) Log.w(TAG, "getHomeFeed: toPost() returned null for doc ${doc.id}")
                post
            }
            
            // Batch check which posts are liked by current user
            val likedPostIds = if (postsWithIds.isNotEmpty()) {
                batchCheckLikes(postsWithIds.map { it.id }, currentUserId)
            } else {
                emptySet()
            }
            
            // Update posts with correct isLikedByMe value
            val list = postsWithIds.map { post ->
                post.copy(isLikedByMe = likedPostIds.contains(post.id))
            }
            
            Log.d(TAG, "getHomeFeed: parsed list.size=${list.size} (raw=$rawCount), liked=${likedPostIds.size}")
            if (snapshot.documents.isNotEmpty()) {
                lastDocumentSnapshot = snapshot.documents.last()
            }
            list
        } catch (e: Exception) {
            Log.e(TAG, "getHomeFeed failed: page=$page", e)
            throw e
        }
    }

    override suspend fun refreshFeed(): List<Post> {
        Log.d(TAG, "refreshFeed: clearing cursor, fetching page 0")
        lastDocumentSnapshot = null
        return getHomeFeed(0, DEFAULT_PAGE_SIZE)
    }

    /**
     * Batch check which posts are liked by the current user.
     * Uses parallel coroutines for efficient batch fetching.
     */
    private suspend fun batchCheckLikes(postIds: List<String>, userId: String): Set<String> = coroutineScope {
        if (postIds.isEmpty()) return@coroutineScope emptySet()
        
        try {
            // Use coroutines to check all posts in parallel
            val results = postIds.map { postId ->
                async {
                    try {
                        val likeDoc = firestore
                            .collection("posts")
                            .document(postId)
                            .collection("likes")
                            .document(userId)
                            .get()
                            .await()
                        if (likeDoc.exists()) postId else null
                    } catch (e: Exception) {
                        Log.e(TAG, "batchCheckLikes: failed for postId=$postId", e)
                        null
                    }
                }
            }
            
            // Wait for all checks to complete and collect results
            val likedPostIds = results.mapNotNull { it.await() }.toSet()
            
            Log.d(TAG, "batchCheckLikes: checked ${postIds.size} posts, found ${likedPostIds.size} liked")
            likedPostIds
        } catch (e: Exception) {
            Log.e(TAG, "batchCheckLikes: failed", e)
            emptySet() // Return empty set on error, posts will show as not liked
        }
    }

    private fun DocumentSnapshot.toPost(isLikedByMe: Boolean = false): Post? {
        return try {
            val data = data ?: return null
            val id = getString("id") ?: this.id
            val authorId = getString("authorId") ?: return null
            val authorName = getString("authorName") ?: return null
            val authorUsername = getString("authorUsername") ?: return null
            val authorProfileImageUrl = getString("authorProfileImageUrl")
            val authorRoleStr = getString("authorRole") ?: "farmer"
            val authorVerificationStatusStr = getString("authorVerificationStatus") ?: "unverified"
            val typeStr = getString("type") ?: "normal"
            val text = getString("text") ?: ""

            val mediaList = (get("media") as? List<Map<String, Any>>)?.mapNotNull { mediaMap ->
                val url = mediaMap["url"] as? String ?: return@mapNotNull null
                val typeStrMedia = mediaMap["type"] as? String ?: "image"
                val thumbnailUrl = mediaMap["thumbnailUrl"] as? String
                PostMedia(
                    url = url,
                    type = when (typeStrMedia) {
                        "video" -> MediaType.VIDEO
                        else -> MediaType.IMAGE
                    },
                    thumbnailUrl = thumbnailUrl?.takeIf { it.isNotEmpty() }
                )
            } ?: emptyList()

            val voiceCaptionMap = get("voiceCaption") as? Map<String, Any>
            val voiceCaption = voiceCaptionMap?.let {
                VoiceContent(
                    url = it["url"] as? String ?: return@let null,
                    durationSeconds = (it["durationSeconds"] as? Number)?.toInt() ?: 0
                )
            }

            val crops = (get("crops") as? List<*>)?.mapNotNull { it as? String } ?: emptyList()
            val hashtags = (get("hashtags") as? List<*>)?.mapNotNull { it as? String } ?: emptyList()

            val locationMap = get("location") as? Map<String, Any>
            val location = locationMap?.let {
                val geoPoint = it["geoPoint"] as? GeoPoint
                PostLocation(
                    name = it["name"] as? String ?: "",
                    latitude = geoPoint?.latitude,
                    longitude = geoPoint?.longitude
                )
            }

            val questionMap = get("question") as? Map<String, Any>
            val question = questionMap?.let {
                QuestionData(
                    targetExpertise = (it["targetExpertise"] as? List<*>)?.mapNotNull { e -> e as? String } ?: emptyList(),
                    targetExpertIds = (it["targetExpertIds"] as? List<*>)?.mapNotNull { e -> e as? String } ?: emptyList(),
                    targetExperts = (it["targetExperts"] as? List<*>)?.mapNotNull { e ->
                        val expertMap = e as? Map<String, Any>
                        expertMap?.let {
                            UserInfo(
                                id = it["id"] as? String ?: "",
                                name = it["name"] as? String ?: "",
                                username = it["username"] as? String ?: "",
                                profileImageUrl = it["profileImageUrl"] as? String,
                                role = stringToUserRole(it["role"] as? String),
                                verificationStatus = stringToVerificationStatus(it["verificationStatus"] as? String)
                            )
                        }
                    } ?: emptyList(),
                    isAnswered = it["isAnswered"] as? Boolean ?: false,
                    bestAnswerCommentId = it["bestAnswerCommentId"] as? String
                )
            }

            val likesCount = (getLong("likesCount") ?: 0L).toInt()
            val commentsCount = (getLong("commentsCount") ?: 0L).toInt()
            val savesCount = (getLong("savesCount") ?: 0L).toInt()
            val createdAt = getTimestamp("createdAt")?.toDate()?.time ?: System.currentTimeMillis()
            val updatedAt = getTimestamp("updatedAt")?.toDate()?.time

            Post(
                id = id,
                authorId = authorId,
                authorName = authorName,
                authorUsername = authorUsername,
                authorProfileImageUrl = authorProfileImageUrl,
                authorRole = stringToUserRole(authorRoleStr),
                authorVerificationStatus = stringToVerificationStatus(authorVerificationStatusStr),
                type = when (typeStr) {
                    "question" -> PostType.QUESTION
                    else -> PostType.NORMAL
                },
                text = text,
                media = mediaList,
                voiceCaption = voiceCaption,
                crops = crops,
                hashtags = hashtags,
                location = location,
                question = question,
                likesCount = likesCount,
                commentsCount = commentsCount,
                savesCount = savesCount,
                isLikedByMe = isLikedByMe,
                isSavedByMe = false, // TODO: Check saved posts similarly
                createdAt = createdAt,
                updatedAt = updatedAt
            )
        } catch (e: Exception) {
            Log.e(TAG, "toPost: Failed to parse feed document", e)
            null
        }
    }

    private fun stringToUserRole(roleStr: String?): UserRole = when (roleStr) {
        "expert" -> UserRole.EXPERT
        "agripreneur" -> UserRole.AGRIPRENEUR
        "input_seller" -> UserRole.INPUT_SELLER
        "agri_lover" -> UserRole.AGRI_LOVER
        else -> UserRole.FARMER
    }

    private fun stringToVerificationStatus(statusStr: String?): VerificationStatus = when (statusStr) {
        "pending" -> VerificationStatus.PENDING
        "verified" -> VerificationStatus.VERIFIED
        "rejected" -> VerificationStatus.REJECTED
        else -> VerificationStatus.UNVERIFIED
    }

    companion object {
        private const val TAG = "FirestoreFeedRepo"
        private const val DATABASE_NAME = "kissangram"
        private const val DEFAULT_PAGE_SIZE = 20
    }
}
